direct_failure_detector/failure_detector.cc

```diff
@@ -0,0 +1,699 @@
+
+            if (it->second == update) {
+                // Safe to remove the entry.
+                _endpoint_updates.erase(it);
+            } else {
+                // While we were updating the endpoint, the user requested the opposite update.
+                // Need to handle this endpoint again.
+            }
+
+            continue;
+        } catch (...) {
+            logger.error("update_endpoint_fiber: failed to add/remove endpoint {}: {}", ep, std::current_exception());
+        }
+
+        // There was an exception.
+        // Wait for a while before proceeding (although the operation should only fail due to OOM, so we're probably already doomed).
+        // Note: `add_endpoint` and `remove_endpoint` provide strong exception guarantees, so we can retry the operation on the same endpoint.
+        try {
+            // Use a dummy abort source for the sleep.
+            // We don't react to shutdowns so the sleep can theoretically prolong it a bit if it happens,
+            // but this codepath should in practice never be reached in the first place, so whatever.
+            abort_source as;
+            co_await _clock.sleep_until(_clock.now() + 10 * _ping_period, as);
+        } catch (sleep_aborted&) {}
+    }
+}
+
+future<> failure_detector::impl::add_endpoint(pinger::endpoint_id ep) {
+    assert(this_shard_id() == 0);
+
+    if (_workers.contains(ep)) {
+        co_return;
+    }
+
+    // Pick a shard with the smallest number of workers to create a new worker.
+    auto shard = std::distance(_num_workers.begin(), std::min_element(_num_workers.begin(), _num_workers.end()));
+    assert(_num_workers.size() == smp::count);
+
+    ++_num_workers[shard];
+    auto [it, _] = _workers.emplace(ep, shard);
+
+    try {
+        co_await _parent.container().invoke_on(shard, [ep] (failure_detector& fd) { fd._impl->create_worker(ep); });
+    } catch (...) {
+        --_num_workers[shard];
+        _workers.erase(it);
+        throw;
+    }
+}
+
+future<> failure_detector::impl::remove_endpoint(pinger::endpoint_id ep) {
+    assert(this_shard_id() == 0);
+
+    auto it = _workers.find(ep);
+    if (it == _workers.end()) {
+        co_return;
+    }
+
+    auto shard = it->second;
+    co_await _parent.container().invoke_on(shard, [ep] (failure_detector& fd) { return fd._impl->destroy_worker(ep); });
+
+    assert(_num_workers.size() == smp::count);
+    assert(shard < _num_workers.size());
+    --_num_workers[shard];
+    _workers.erase(it);
+
+    // Note: removing endpoints may create imbalance of worker distribution across shards.
+    // Right now we don't do anything with it, as we don't expect huge number of workers,
+    // and if new workers are added eventually, balance will be restored.
+    // Alternatively we could migrate running workers among shards but it's probably not worth it.
+}
+
+void failure_detector::impl::create_worker(pinger::endpoint_id ep) {
+    // To provide strong exception guarantee.
```

Comment:
What does this comment means? That you need to provide it or that the guard is needed for strong guaranty?

Reply:
That the guard is used to provide it

Reply:
OK. So I assume the fact that `emplace_back` below can throw is not a problem?

Reply:
It is not, it's emplacing into `guards`, not changing any state inside the failure detector